### PR TITLE
[GH-726] NG mining

### LINKED ISSUE #726: Adjust candidate block generation/mining
Key block candidates should be generated and mined while the node isn't the current leader, micro blocks should be produced when the node is leading

### LINKED ISSUE #776: Adjust genesis block
Should be a key block

### PR DESCRIPTION
**Pull request by d-velev**
closes #726 closes #776 
This also adjusts the genesis block.
things to be done later on: #736 #727 and tests to be refactored when everything needed is added

### CODE REVIEW

**Comment by thepiwo** [comment]
With what issues will this be fixed? If these are actual "features" its okay to do later, if this is within the scope of this feature it should be fixed here

**Comment by d-velev** [comment]
@thepiwo block validation and miner reward calculation adjustments are a separate issue, also tests need to be redone in some way so that transactions pass without any balance, I think this is a minimal solution that doesn't change too many things, with more of the issues resolved at some point I'll probably do another issue that fixes tests and removes all old code

**Comment by thepiwo** [comment]
@d-velev could you reference these issues here, so everyone can see

**Review by cytadela8 — COMMENTED** [review]
Just a question.

File: apps/aecore/lib/aecore/chain/chainstate.ex [hunk 1]
```diff
@@ -64,27 +64,29 @@ defmodule Aecore.Chain.Chainstate do
...
-          {:ok, updated_chainstate} ->
-            {:cont, updated_chainstate}
-
-          {:error, reason} ->
-            {:halt, reason}
-        end
-      end)
-
-    final_chainstate =
-      calculate_miner_reward_chain_state(txs, updated_chainstate, block_height, miner)
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply apps/aecore/lib/aecore/chain/chainstate.ex]: Ok.

**Review by cytadela8 — APPROVED** [review]

File: apps/aecore/lib/aecore/chain/worker.ex [hunk 2]
```diff
@@ -224,31 +243,29 @@ defmodule Aecore.Chain.Worker do
     end
   end
 
-  @spec add_block(Block.t()) :: :ok | {:error, String.t()}
-  def add_block(%Block{header: %Header{prev_hash: prev_hash}} = block) do
+  @spec add_block(Block.t(), boolean()) :: :ok | {:error, String.t()}
+  def add_block(
+        %{header: %{prev_hash: prev_hash, prev_key_hash: prev_key_hash, height: height}} = block,
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/chain/worker.ex]: No type safety here. Add "when block.__struct__ in [KeyBlock, MicroBlock]"
The typespec is wrong.

File: apps/aecore/lib/aecore/chain/worker.ex [hunk 3]
```diff
@@ -224,31 +243,29 @@ defmodule Aecore.Chain.Worker do
...
+      add_validated_block(block, new_chain_state, loop_micro_block)
     else
       err -> err
     end
   end
 
-  @spec add_validated_block(Block.t(), Chainstate.t()) :: :ok
-  defp add_validated_block(%Block{} = block, chain_state) do
-    GenServer.call(__MODULE__, {:add_validated_block, block, chain_state})
+  @spec add_validated_block(Block.t(), Chainstate.t(), boolean()) :: :ok
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/chain/worker.ex]: Typespec

File: apps/aecore/lib/aecore/chain/worker.ex [hunk 4]
```diff
@@ -338,19 +355,47 @@ defmodule Aecore.Chain.Worker do
 
   def handle_call(
         {:add_validated_block,
-         %Block{
-           header: %Header{prev_hash: prev_hash, height: height, target: target} = header,
-           txs: txs
-         } = new_block, new_chain_state},
+         %{
+           header: %{prev_hash: prev_hash, height: height, time: time} = header
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/chain/worker.ex]: Add type guard check.

**Review by gorbak25 — CHANGES_REQUESTED** [review]
Partial review.

File: apps/aecore/lib/aecore/chain/worker.ex (hunk 2 above)
Comment (d-velev) [reply apps/aecore/lib/aecore/chain/worker.ex]: I don't think there's a need for explicit type checking in this case, we do checks in other functions/fragments that are part of this

File: apps/aecore/lib/aecore/util/header.ex [hunk 5]
```diff
@@ -0,0 +1,28 @@
+defmodule Aecore.Util.Header do
+  @moduledoc """
+  Header utilities
+  """
+  alias Aecore.Chain.KeyHeader
+  alias Aecore.Chain.MicroHeader
+
+  @spec hash(KeyHeader.t() | MicroHeader.t()) :: binary()
+  def hash(header) do
+    case header do
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/util/header.ex]: This case can be removed by specifying two clauses in the hash function, eg:
```
def hash(%KeyHeader{}) do
...
end
def hash(%MicroHeader()) do
...
end
...
```
Or by writing:
 ``` 
def hash(header) when header.__struct__ in [KeyHeader, MicroHeader] do
    header.__struct__.hash(header)
end
```

File: apps/aecore/lib/aecore/util/header.ex [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      %KeyHeader{} ->
+        KeyHeader.hash(header)
+
+      %MicroHeader{} ->
+        MicroHeader.hash(header)
+    end
+  end
+
+  def top_key_block_hash(prev_header) do
+    case prev_header do
```
Comment (grzegorz225) [inline apps/aecore/lib/aecore/util/header.ex]: Same as above

**Review by gorbak25 — COMMENTED** [review]
Looks good.

File: apps/aecore/lib/aecore/chain/chainstate.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (thepiwo) [reply apps/aecore/lib/aecore/chain/chainstate.ex]: what issue is it, can you reference here for all to see?

File: apps/aecore/lib/aecore/util/header.ex (hunk 5 above)
Comment (d-velev) [reply apps/aecore/lib/aecore/util/header.ex]: Code reduction is always nice, will change it 😆 I dislike the second one

**Review by gorbak25 — CHANGES_REQUESTED** [review]
Could you make sure that the code compiles without warnings?
https://travis-ci.org/aeternity/elixir-node/builds/455887419#L600

**Comment by cytadela8** [comment]
@grzegorz225 This is "In progress" code. A lot of work is done here and some warnings are due to unimplemented features. (They are warning for example that something isn't used were it should be. I think proper way of fixing such warning is by implementing NG)

**Review by gorbak25 — APPROVED** [review]

**Review by DanielaIvanova — APPROVED** [review]

**Pull request merged**